### ISSUE TITLE
Create standardized way of rendering parts of the page

### ISSUE THREAD

**Issue by lauriii** [jsdrupal/drupal-admin-ui#184]
It would be useful to have standardized way of providing data from page level components so that it would be possible to render important parts of the page in a standardized way. Good example of this is the page title which we probably want always to exist and we want it to be always rendered the same way. There is probably other use cases as well, for example local actions and tasks.

One of the methods would be to use interfaces, which unfortunately doesn't exist in JavaScript. Using TypeScript would solve this #64.

Other option would be to use https://github.com/gaearon/react-side-effect library which allows components to trigger global side effects.

**Comment by justafish** [comment]
Could you clarify a bit further what you mean here? I'm not understanding what the requirements are that aren't solved by having e.g. a PageTitle component.

**Comment by lauriii** [comment]
Mainly this is about the fact that we want to always have the page title rendered in the same place on the page. If rendering title is done manually by components rendering pages, that might lead into inconsistency between pages. This is a pre-existing UI pattern in Drupal which we might want to bring into this application.

**Comment by dawehner** [comment]
I guess a requirement we have is not only that we want the page title to always appear in the same place, it is more about there is always one.

Is the admin UX design team 100% sure that this is the pattern which will exist in the future?

On the implementation side of things I am wondering whether we could use the fact that react components can return not just elements, but also array of elements.

**Comment by ckrina** [comment]
In the title situation specifically, there will be always a title but I'm not sure it will always behave the same way. In example, let's say you are creating a new node: now the title is something like "Create Basic page" but in the future it could be changed to something that represents the real title for new content you are creating.
And local actions could be extended by other modules maybe?
Maybe the link to the user profile could be a recurrent thing that doesn't change.

**Comment by lauriii** [comment]
> Maybe the link to the user profile could be a recurrent thing that doesn't change.

This seems like something that doesn't change per page.

> In the title situation specifically, there will be always a title but I'm not sure it will always behave the same way. In example, let's say you are creating a new node: now the title is something like "Create Basic page" but in the future it could be changed to something that represents the real title for new content you are creating.

The question is that do we want to force every page to have a title as a UX pattern or not.

**Comment by ckrina** [comment]
Yes, there should be a title to let people have context. But it might not look the same, be placed on the  same region or behave the same.
